File: backend/tavus_service.py
"""
Tavus CVI Service — persona creation, conversation lifecycle, cleanup.
Docs: https://docs.tavus.io/api-reference
"""

import logging
import os

import httpx

logger = logging.getLogger(__name__)

TAVUS_BASE_URL = "https://tavusapi.com/v2"
_TIMEOUT = httpx.Timeout(30.0, connect=10.0)

# ---------------------------------------------------------------------------
# System prompt template (injected into Tavus persona)
# ---------------------------------------------------------------------------
_SYSTEM_PROMPT_TEMPLATE = """
You are InterviewBot, a professional and empathetic AI interview coach.
You are conducting a {interview_type} interview at {difficulty} level in {language}.

Candidate Resume:
- Name: {name}
- Skills: {skills}
- Experience: {experience}
- Companies: {companies}
- Education: {education}

Your Questions (ask in this order, adapt follow-ups based on answers):
{numbered_questions}

Behavior Rules:
- Start with: "Hello {name}! Welcome to your mock interview. I'm InterviewBot. Let's begin with a quick introduction — please tell me about yourself."
- Ask ONE question at a time. Never ask two questions together.
- After each answer, give a 1-sentence acknowledgment, then ask the next question.
- If the candidate's answer is under 15 words, prompt once: "Could you elaborate a bit more on that?"
- After all 7 questions, say: "That brings us to the end of the interview. You did great, {name}. Your detailed performance report will be ready shortly. Thank you!"
- Never reveal scores, never give feedback during the interview.
- Be warm, professional, and encouraging throughout.
- Do not go off-topic. Bring the conversation back to the interview if needed.
""".strip()


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _is_test_mode() -> bool:
    """Return True when running with test credentials or forced test mode."""
    val = os.getenv("TAVUS_API_KEY", "").strip().lower()
    return val == "test" or val == ""


def _get_headers() -> dict:
    api_key = os.getenv("TAVUS_API_KEY", "")
    if not api_key:
        raise EnvironmentError("TAVUS_API_KEY is not set in the environment.")
    return {
        "x-api-key": api_key,
        "Content-Type": "application/json",
    }


def _build_system_prompt(
    resume_data: dict,
    questions: list[str],
    interview_type: str,
    difficulty: str,
    language: str,
) -> str:
    numbered_questions = "\n".join(
        f"{i + 1}. {q}" for i, q in enumerate(questions)
    )
    return _SYSTEM_PROMPT_TEMPLATE.format(
        interview_type=interview_type,
        difficulty=difficulty,
        language=language,
        name=resume_data.get("name", "Candidate"),
        skills=", ".join(resume_data.get("skills", [])) or "Not specified",
        experience=f"{resume_data.get('total_experience_years', 0)} years",
        companies=", ".join(resume_data.get("companies", [])) or "Not specified",
        education=", ".join(resume_data.get("education", [])) or "Not specified",
        numbered_questions=numbered_questions,
    )


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

async def create_persona(
    resume_data: dict,
    questions: list[str],
    interview_type: str,
    difficulty: str,
    language: str,
) -> str:
    """
    Create a Tavus persona injected with the candidate's resume and interview questions.

    Returns:
        persona_id (str)

    Raises:
        RuntimeError: on Tavus API error
        EnvironmentError: if TAVUS_API_KEY is missing
    """
    system_prompt = _build_system_prompt(
        resume_data, questions, interview_type, difficulty, language
    )
    replica_id = os.getenv("TAVUS_REPLICA_ID", "rf4e9d9790f0")

    payload = {
        "persona_name": f"InterviewBot-{interview_type}-{difficulty}",
        "system_prompt": system_prompt,
        "default_replica_id": replica_id,
        "pipeline_mode": "full",
        "context": (
            f"This is a {difficulty}-level {interview_type} interview. "
            f"The candidate is {resume_data.get('name', 'the candidate')}."
        ),
        "layers": {
            "llm": {
                "model": "tavus-gpt-4.1",
            }
        },
    }

    logger.info("Creating Tavus persona for %s ...", resume_data.get("name", "candidate"))

    # ── TEST MODE ──────────────────────────────────────────────────────────────
    if _is_test_mode():
        logger.warning("TEST MODE: returning mock persona_id (no real Tavus call)")
        return "test-persona-123"

    async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
        try:
            resp = await client.post(
                f"{TAVUS_BASE_URL}/personas",
                json=payload,
                headers=_get_headers(),
            )
            resp.raise_for_status()
        except httpx.HTTPStatusError as exc:
            body = exc.response.text
            logger.error("Tavus persona creation failed [%s]: %s", exc.response.status_code, body)
            raise RuntimeError(
                f"Tavus persona creation failed ({exc.response.status_code}): {body}"
            ) from exc
        except httpx.RequestError as exc:
            logger.error("Network error reaching Tavus API: %s", exc)
            raise RuntimeError(f"Network error reaching Tavus: {exc}") from exc

    data = resp.json()
    persona_id = data.get("persona_id") or data.get("id")
    if not persona_id:
        raise RuntimeError(f"Tavus returned no persona_id. Response: {data}")

    logger.info("Tavus persona created: %s", persona_id)
    return persona_id


async def create_conversation(
    persona_id: str,
    candidate_name: str,
    session_id: str,
) -> dict:
    """
    Start a Tavus CVI conversation for the given persona.

    Returns:
        dict with keys: conversation_id, conversation_url

    Raises:
        RuntimeError: on Tavus API error
    """
    backend_url = os.getenv("BACKEND_URL", "http://localhost:8000")
    callback_url = f"{backend_url}/api/tavus-webhook"

    payload = {
        "persona_id": persona_id,
        "conversation_name": f"Interview: {candidate_name}",
        "callback_url": callback_url,
        "properties": {
            "max_call_duration": 1800,  # 30 minutes hard cap
            "participant_left_timeout": 60,
            "enable_recording": False,
            "apply_greenscreen": False,
        },
        "custom_greeting": (
            f"Hello {candidate_name}! Welcome to your mock interview. "
            "I'm InterviewBot. Let's begin with a quick introduction — "
            "please tell me about yourself."
        ),
    }

    logger.info("Creating Tavus conversation for persona %s ...", persona_id)

    # ── TEST MODE ──────────────────────────────────────────────────────────────
    if _is_test_mode():
        logger.warning("TEST MODE: returning mock conversation URL (no real Tavus call)")
        name_slug = candidate_name.lower().replace(" ", "-")
        return {
            "conversation_id": f"test-conv-{name_slug}-123",
            "conversation_url": "https://tavus.daily.co/test-room",
        }

    async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
        try:
            resp = await client.post(
                f"{TAVUS_BASE_URL}/conversations",
                json=payload,
                headers=_get_headers(),
            )
            resp.raise_for_status()
        except httpx.HTTPStatusError as exc:
            body = exc.response.text
            logger.error(
                "Tavus conversation creation failed [%s]: %s",
                exc.response.status_code,
                body,
            )
            raise RuntimeError(
                f"Tavus conversation creation failed ({exc.response.status_code}): {body}"
            ) from exc
        except httpx.RequestError as exc:
            logger.error("Network error reaching Tavus API: %s", exc)
            raise RuntimeError(f"Network error reaching Tavus: {exc}") from exc

    data = resp.json()
    conversation_id = data.get("conversation_id") or data.get("id")
    conversation_url = data.get("conversation_url")

    if not conversation_id or not conversation_url:
        raise RuntimeError(
            f"Tavus returned incomplete conversation data. Response: {data}"
        )

    logger.info(
        "Tavus conversation created: id=%s url=%s", conversation_id, conversation_url
    )
    return {
        "conversation_id": conversation_id,
        "conversation_url": conversation_url,
    }


async def end_conversation(conversation_id: str) -> None:
    """
    Terminate an active Tavus conversation via DELETE.

    Raises:
        RuntimeError: on Tavus API error
    """
    logger.info("Ending Tavus conversation: %s", conversation_id)

    # ── TEST MODE ──────────────────────────────────────────────────────────────
    if _is_test_mode():
        logger.warning("TEST MODE: skipping real Tavus end_conversation call")
        return

    async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
        try:
            resp = await client.delete(
                f"{TAVUS_BASE_URL}/conversations/{conversation_id}",
                headers=_get_headers(),
            )
            resp.raise_for_status()
        except httpx.HTTPStatusError as exc:
            body = exc.response.text
            logger.error(
                "Tavus conversation end failed [%s]: %s",
                exc.response.status_code,
                body,
            )
            raise RuntimeError(
                f"Tavus conversation end failed ({exc.response.status_code}): {body}"
            ) from exc
        except httpx.RequestError as exc:
            logger.error("Network error reaching Tavus API: %s", exc)
            raise RuntimeError(f"Network error reaching Tavus: {exc}") from exc

    logger.info("Tavus conversation %s ended successfully.", conversation_id)


async def get_conversation_status(conversation_id: str) -> dict:
    """
    Fetch current status of a Tavus conversation.

    Returns:
        Raw Tavus conversation object dict.
    """
    async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
        try:
            resp = await client.get(
                f"{TAVUS_BASE_URL}/conversations/{conversation_id}",
                headers=_get_headers(),
            )
            resp.raise_for_status()
        except httpx.HTTPStatusError as exc:
            raise RuntimeError(
                f"Tavus status check failed ({exc.response.status_code}): {exc.response.text}"
            ) from exc
    return resp.json()

File: backend-java/src/main/java/com/example/mockmate/service/TavusService.java
package com.example.mockmate.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.annotation.PostConstruct;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;
import org.springframework.web.reactive.function.client.WebClientResponseException;
import reactor.core.publisher.Mono;

import java.time.Duration;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Slf4j
@Service
public class TavusService {

    private final WebClient webClient;
    private final ObjectMapper objectMapper;

    @Value("${tavus.api-key}")
    private String tavusApiKey;

    @Value("${tavus.replica-id}")
    private String tavusReplicaId;

    @PostConstruct
    public void logConfig() {
        boolean live = !isTestMode();
        log.info("[TavusService] Mode: {} | Replica: {} | Key prefix: {}",
                live ? "LIVE" : "TEST",
                tavusReplicaId,
                tavusApiKey.length() > 8 ? tavusApiKey.substring(0, 8) + "..." : "(short/empty)");
    }

    @Value("${backend.url:http://localhost:8080}")
    private String backendUrl;

    private static final String SYSTEM_PROMPT_TEMPLATE = """
            You are InterviewBot, a professional and empathetic AI interview coach.
            You are conducting a %s interview at %s level in %s.
            
            Candidate Resume:
            - Name: %s
            - Skills: %s
            - Experience: %s
            - Companies: %s
            - Education: %s
            
            Your Questions (ask in this order, adapt follow-ups based on answers):
            %s
            
            Behavior Rules:
            - Start with: "Hello %s! Welcome to your mock interview. I'm InterviewBot. Let's begin with a quick introduction — please tell me about yourself."
            - Ask ONE question at a time. Never ask two questions together.
            - After each answer, give a 1-sentence acknowledgment, then ask the next question.
            - If the candidate's answer is under 15 words, prompt once: "Could you elaborate a bit more on that?"
            - After all 7 questions, say: "That brings us to the end of the interview. You did great, %s. Your detailed performance report will be ready shortly. Thank you!"
            - Never reveal scores, never give feedback during the interview.
            - Be warm, professional, and encouraging throughout.
            - Do not go off-topic. Bring the conversation back to the interview if needed.
            """;

    public TavusService(WebClient.Builder webClientBuilder, ObjectMapper objectMapper) {
        this.webClient = webClientBuilder.baseUrl("https://tavusapi.com/v2").build();
        this.objectMapper = objectMapper;
    }

    private boolean isTestMode() {
        return tavusApiKey == null
                || tavusApiKey.isBlank()
                || "test".equalsIgnoreCase(tavusApiKey.trim());
    }

    public String createPersona(Map<String, Object> resumeData, List<String> questions, String interviewType, String difficulty, String language) {
        StringBuilder numberedQuestions = new StringBuilder();
        for (int i = 0; i < questions.size(); i++) {
            numberedQuestions.append(i + 1).append(". ").append(questions.get(i)).append("\n");
        }

        String name = (String) resumeData.getOrDefault("name", "Candidate");
        String skills = String.join(", ", (List<String>) resumeData.getOrDefault("skills", List.of()));
        if (skills.isEmpty()) skills = "Not specified";
        String experience = resumeData.getOrDefault("totalExperienceYears", 0) + " years";
        String companies = String.join(", ", (List<String>) resumeData.getOrDefault("companies", List.of()));
        if (companies.isEmpty()) companies = "Not specified";
        String education = String.join(", ", (List<String>) resumeData.getOrDefault("education", List.of()));
        if (education.isEmpty()) education = "Not specified";

        String systemPrompt = String.format(SYSTEM_PROMPT_TEMPLATE,
                interviewType, difficulty, language,
                name, skills, experience, companies, education,
                numberedQuestions.toString(),
                name, name
        );

        Map<String, Object> payload = new HashMap<>();
        payload.put("persona_name", "InterviewBot-" + interviewType + "-" + difficulty);
        payload.put("system_prompt", systemPrompt);
        payload.put("default_replica_id", tavusReplicaId);
        payload.put("context", "This is a " + difficulty + "-level " + interviewType + " interview. The candidate is " + name + ".");
        // NOTE: Do NOT include 'layers.llm.model' or 'pipeline_mode' unless on an Enterprise plan
        //       They cause 422/400 errors that silently fall through to mock mode.

        if (isTestMode()) {
            log.warn("TEST MODE: returning mock persona_id");
            return "test-persona-123";
        }

        try {
            String response = webClient.post()
                    .uri("/personas")
                    .header("x-api-key", tavusApiKey.trim())
                    .header("Content-Type", "application/json")
                    .bodyValue(payload)
                    .retrieve()
                    .onStatus(status -> status.is4xxClientError() || status.is5xxServerError(),
                        clientResponse -> clientResponse.bodyToMono(String.class)
                            .flatMap(errorBody -> {
                                log.error("[TavusService] Persona HTTP {} — body: {}", clientResponse.statusCode(), errorBody);
                                return Mono.error(new RuntimeException("Tavus persona API error: " + errorBody));
                            })
                    )
                    .bodyToMono(String.class)
                    .timeout(Duration.ofSeconds(30))
                    .block();
            log.info("[TavusService] Persona API raw response: {}", response);
            JsonNode root = objectMapper.readTree(response);
            if (root.has("persona_id")) return root.get("persona_id").asText();
            if (root.has("id"))         return root.get("id").asText();
            log.error("[TavusService] Persona response had no ID. Full response: {}", response);
            throw new RuntimeException("Tavus returned no persona_id. Response: " + response);
        } catch (WebClientResponseException e) {
            log.error("[TavusService] Persona HTTP {} — body: {}", e.getStatusCode(), e.getResponseBodyAsString());
            return "test-persona-123";
        } catch (Exception e) {
            log.error("[TavusService] Persona creation error: {}", e.getMessage());
            return "test-persona-123";
        }
    }

    public Map<String, String> createConversation(String personaId, String candidateName, String sessionId) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("persona_id", personaId);
        payload.put("conversation_name", "Interview: " + candidateName);
        
        String callbackUrl = backendUrl + "/api/tavus-webhook";
        if (callbackUrl != null && !callbackUrl.contains("localhost") && !callbackUrl.contains("127.0.0.1") && callbackUrl.startsWith("https://")) {
            payload.put("callback_url", callbackUrl);
            log.info("[TavusService] Configured webhook callback_url: {}", callbackUrl);
        } else {
            log.warn("[TavusService] Skipping callback_url because backend URL is local or non-HTTPS: {}", callbackUrl);
        }

        payload.put("properties", Map.of(
                "max_call_duration", 1800,
                "participant_left_timeout", 60,
                "enable_recording", false,
                "apply_greenscreen", false
        ));
        payload.put("custom_greeting", "Hello " + candidateName + "! Welcome to your mock interview. I'm InterviewBot. Let's begin with a quick introduction — please tell me about yourself.");

        if (isTestMode()) {
            log.warn("TEST MODE: returning mock conversation URL");
            return Map.of(
                    "conversation_id", "test-conv-" + candidateName.toLowerCase().replace(" ", "-") + "-123",
                    "conversation_url", "https://tavus.daily.co/test-room"
            );
        }

        try {
            String response = webClient.post()
                    .uri("/conversations")
                    .header("x-api-key", tavusApiKey.trim())
                    .header("Content-Type", "application/json")
                    .bodyValue(payload)
                    .retrieve()
                    .onStatus(status -> status.is4xxClientError() || status.is5xxServerError(),
                        clientResponse -> clientResponse.bodyToMono(String.class)
                            .flatMap(errorBody -> {
                                log.error("[TavusService] Conversation HTTP {} — body: {}", clientResponse.statusCode(), errorBody);
                                return Mono.error(new RuntimeException("Tavus conversation API error: " + errorBody));
                            })
                    )
                    .bodyToMono(String.class)
                    .timeout(Duration.ofSeconds(30))
                    .block();
            log.info("[TavusService] Conversation API raw response: {}", response);
            JsonNode root = objectMapper.readTree(response);
            String conversationId  = root.has("conversation_id") ? root.get("conversation_id").asText() : root.get("id").asText();
            String conversationUrl = root.get("conversation_url").asText();
            log.info("[TavusService] Live conversation → ID: {}, URL: {}", conversationId, conversationUrl);
            return Map.of("conversation_id", conversationId, "conversation_url", conversationUrl);
        } catch (WebClientResponseException e) {
            log.error("[TavusService] Conversation HTTP {} — body: {}", e.getStatusCode(), e.getResponseBodyAsString());
            return Map.of(
                    "conversation_id", "test-conv-fallback",
                    "conversation_url", "https://tavus.daily.co/test-room"
            );
        } catch (Exception e) {
            log.error("[TavusService] Conversation creation error: {}", e.getMessage());
            return Map.of(
                    "conversation_id", "test-conv-fallback",
                    "conversation_url", "https://tavus.daily.co/test-room"
            );
        }
    }

    public void endConversation(String conversationId) {
        if (isTestMode()) return;
        try {
            webClient.post()
                    .uri("/conversations/" + conversationId + "/end")
                    .header("x-api-key", tavusApiKey)
                    .retrieve()
                    .bodyToMono(Void.class)
                    .block();
            log.info("Successfully ended Tavus conversation: {}", conversationId);
        } catch (Exception e) {
            log.error("Failed to end Tavus conversation: {}", conversationId, e);
        }
    }

    public List<Map<String, Object>> getTranscript(String conversationId) {
        if (isTestMode()) {
            log.warn("TEST MODE: returning empty mock transcript");
            return List.of();
        }

        try {
            String response = webClient.get()
                    .uri("/conversations/" + conversationId + "?verbose=true")
                    .header("x-api-key", tavusApiKey)
                    .retrieve()
                    .bodyToMono(String.class)
                    .block();

            JsonNode root = objectMapper.readTree(response);
            log.info("Fetched Tavus conversation data for {}", conversationId);

            List<Map<String, Object>> turns = new ArrayList<>();
            JsonNode transcriptNode = null;

            if (root.has("transcript")) {
                transcriptNode = root.get("transcript");
            } else if (root.has("events") && root.get("events").isArray()) {
                for (JsonNode eventNode : root.get("events")) {
                    if ("application.transcription_ready".equals(eventNode.path("event_type").asText())) {
                        JsonNode props = eventNode.get("properties");
                        if (props != null && props.has("transcript")) {
                            transcriptNode = props.get("transcript");
                            break;
                        }
                    }
                }
            }

            if (transcriptNode != null && transcriptNode.isArray()) {
                for (JsonNode turnNode : transcriptNode) {
                    Map<String, Object> turn = new HashMap<>();
                    String text = turnNode.has("content") ? turnNode.path("content").asText() : turnNode.path("text").asText();

                    // ── SYSTEM PROMPT LEAK GUARD ────────────────────────────────────────────
                    // Tavus sometimes returns the persona system prompt as the first "assistant"
                    // turn. Detect and drop it: system prompts are extremely long and contain
                    // signature strings that never appear in real interview speech.
                    boolean isSystemPromptLeak = text.contains("You are InterviewBot")
                            || text.contains("Behavior Rules:")
                            || text.contains("SYSTEM_PROMPT")
                            || (text.length() > 2000 && text.contains("interview"));
                    if (isSystemPromptLeak) {
                        log.warn("[TavusService] Dropping system-prompt leak turn ({} chars) for conversation {}",
                                text.length(), conversationId);
                        continue;
                    }
                    // ────────────────────────────────────────────────────────────────────────

                    turn.put("text", text);

                    String role = turnNode.path("role").asText("candidate");
                    turn.put("role", "assistant".equalsIgnoreCase(role) || "ai".equalsIgnoreCase(role) ? "interviewer" : "candidate");
                    
                    double startTime = turnNode.has("start_time") 
                            ? turnNode.path("start_time").asDouble(0.0) 
                            : turnNode.path("seconds_from_start").asDouble(0.0);
                    
                    turn.put("timestamp_ms", System.currentTimeMillis()); // Fallback for now
                    turn.put("relative_seconds", startTime);
                    
                    turns.add(turn);
                }
            }
            return turns;
        } catch (Exception e) {
            log.error("Failed to fetch Tavus transcript", e);
            return List.of();
        }
    }
}
